### PR TITLE
fix goto definition when inbetween tokens

### PR DESCRIPTION
**Pull request by succcubbus**
fixes both goto_definition and goto_type_definition.
before, when running goto between some non-trivia token and an
identifier, goto would be attempted for the non-trivia token.
but this does not make sense for e.g. L_PAREN or COLONCOLON only for
IDENTs.

this resulted in goto actions not working when running them on the first
character of some identifier e.g. for `module::<|>method()` or
`method(<|>parameter)`.

now only IDENTs will be searched for in goto actions, though i'm not sure
if this is correct or if goto should also work for some other token types.

### CODE REVIEW

**Comment by succcubbus** [comment]
i just saw, is should probably also fix this for `hover`

**Comment by succcubbus** [comment]
`hover` is more complex though, because it is actually defined for some of these cases. i'll need to look at it a bit more.

File: crates/ra_ide/src/goto_definition.rs [hunk 1]
```diff
@@ -20,7 +20,7 @@ pub(crate) fn goto_definition(
 ) -> Option<RangeInfo<Vec<NavigationTarget>>> {
     let file = db.parse_or_expand(position.file_id.into())?;
     let original_token =
-        file.token_at_offset(position.offset).filter(|it| !it.kind().is_trivia()).next()?;
+        file.token_at_offset(position.offset).find(|it| it.kind() == SyntaxKind::IDENT)?;
```
Comment (matklad) [inline crates/ra_ide/src/goto_definition.rs]: Hm, does this pass tests? I *think* we have goto def working for tuple fields (`foo.0` syntax), so if tests pass then we miss one!

Comment (matklad) [reply crates/ra_ide/src/goto_definition.rs]: I think ideally this shold work like pick_best in extend selection: https://github.com/rust-analyzer/rust-analyzer/blob/9c9f4635b4eb7d987717b4064a7de789f7983e2a/crates/ra_ide/src/extend_selection.rs#L141

**Comment by succcubbus** [comment]
so, i fixed this for `hover` by just trying either token and using the first that works.
but maybe some heuristic like trying IDENTs first could be useful for users.
possibly this should also be the preferred way for goto.

File: crates/ra_ide/src/goto_definition.rs (hunk 1 above)
Comment (succcubbus) [reply crates/ra_ide/src/goto_definition.rs]: yeah, test pass currently. i'll add a test case for that.

Comment (succcubbus) [reply crates/ra_ide/src/goto_definition.rs]: do you think the solution i used for `hover` i.e. just trying both tokens is acceptable here?

Comment (succcubbus) [reply crates/ra_ide/src/goto_definition.rs]: i see, pick_best seems like a good solution, i'll try implementing that for the gotos and hover

Comment (matklad) [reply crates/ra_ide/src/goto_definition.rs]: No strong opinion here, but like the `pick_test` solution slightly more. i.e, we chose based only on  the kind of the token (but do this properly, prefering identifiers over numbers over punctuation)

**Comment by succcubbus** [comment]
i implemented a `pick_best` for goto and hover.
please note that due to the implementation, if the two tokens have the same priority, the right one will be preferred.

**Comment by matklad** [comment]
bors r+

Thanks!

**Pull request merged**